### DISCUSSION TITLE
h2owave R Package (rdriver) Testing

### DISCUSSION THREAD

**Discussion by ashrith** [h2oai/wave#668] [General]
Hello and thank you for signing up to test `h2owave` package for R. Internally, it is called as rdriver by the wave team. 

To be able to successfully run the `h2owave` library, and example codes please follow the steps below. 

1. There are 5 libraries that are needed by `h2owave` library
    1. `jsonlite`
    2. `devtools`
    3. `httr`
    4. `R6`
    5. `stringr` 
    You can install these packages by conveniently cutting and pasting the following code and by choosing a mirror:
		`install.packages(c("R6","devtools","jsonlite","stringr","httr"))`

2. Once these libraries are installed you can go ahead and install the `h2owave` package. 
    1. This way:
`install.packages(“path_to_package/h2owave_0.0.14.tar.gz",repos=NULL,type="source")`
	2. or from the SHELL this way
`$R CMD INSTALL path_to_package/h2owave_0.0.14.tar.gz`

3. Download the wave server version 0.13 from here https://github.com/h2oai/wave/releases/tag/v0.13.0
4. Untar the wave server, and `cd` into the directory. 
5. After installing the library you will now have to start (if you haven’t already) your wave server. At your SHELL you can run `$./waved -debug`
6. After installing the library you can start using the example applets in two ways:
    1. Line-by-line:
        1. Unzip the examples
        2. Open any one of the seven applets
        3. copy each line of code into the R interpreter environment and run it. 
        4. The functionality of each one of code in each applet is described. 
    2. Source:
        1. in the R interpreter type `> source(“path_to_example/hello_world.R”)`
7. You will be able to see the output of each applet on your browser at `http://localhost:10101/demo`

8. Once you have completed running of the code successfully, Please answer the following:

    1. Was the installation of the library easy? 
    2. Did you understand each of the examples clearly?
        1. Was the code readable?
    3. Did code in the examples make sense to you as a R user? 
        1. Was the code idiomatic?
    4. Were you able to execute the examples easily?
        1. Was the code readable?
        2. Was the code brief?
        3. Was the code ergonomic? 

    1. Any other comments you would like to provide about the package? examples? experience? 

We are making changes to the package as we speak, so if something is missing please do mention it. Especially docs. We would be happy to incorporate your feedback. 

__Updates:__
1. Added `devtools` in the installation list. 
2. Fixed the correct quotes """ enclosing `httr`
3. __updated the `h2owave_0.0.14.tar.gz` (please get the new one here below)__
4. Added specific information about wave server. 

[h2owave_0.0.14.tar.gz](https://github.com/h2oai/wave/files/6201635/h2owave_0.0.14.tar.gz)
[pre-examples.tar.gz](https://github.com/h2oai/wave/files/6198715/pre-examples.tar.gz)

**Comment by grigory93** [comment]
any requirements on R version?

**Reply by ashrith** [reply]
@grigory93 the package is built on `R version 4.0.2 (2020-06-22)`. Let me know if you have an older/newer version and have any compatibility issues.

**Reply by ashrith** [reply]
@grigory93 I updated the library to reflect some changes. Please use the new one provided here.

**Comment by woobe** [comment]
need to fix the final double quotation mark in `install.packages(c("R6","jsonlite","stringr","httr”))`, it is `”` instead of `"`

**Reply by woobe** [reply]
I needed to install ‘devtools’ before step 2.

**Reply by ashrith** [reply]
@woobe thanks for the quotes, and the `devtools` installation. I assumed that everybody would have `devtools` obviously I was wrong.

**Reply by ashrith** [reply]
@woobe updated the guide as per your comments! thanks.

**Reply by grigory93** [reply]
just wondering why do we need `devtools` ?

**Reply by woobe** [reply]
There was an error message about `devtools` and `roxygen2` missing when I tried to install wave the first time (vanilla R 4.0.4 setup in a Virtualbox). I installed `devtools` (which also installed `roxygen2`) and the problem was gone.

**Comment by woobe** [comment]
All examples work and well-commented. I think we need one or two examples of visualization with popular packages like `ggplot2` / `plotly`.

`> sessionInfo()`
`R version 4.0.4 (2021-02-15)`
`Platform: x86_64-pc-linux-gnu (64-bit)`
`Running under: Ubuntu 20.04.2 LTS`

`Matrix products: default`
`BLAS:   /usr/lib/x86_64-linux-gnu/openblas-pthread/libblas.so.3`
`LAPACK: /usr/lib/x86_64-linux-gnu/openblas-pthread/liblapack.so.3`

`locale:`
` [1] LC_CTYPE=en_GB.UTF-8       LC_NUMERIC=C               LC_TIME=en_GB.UTF-8        LC_COLLATE=en_GB.UTF-8    `
` [5] LC_MONETARY=en_GB.UTF-8    LC_MESSAGES=en_GB.UTF-8    LC_PAPER=en_GB.UTF-8       LC_NAME=C                 `
` [9] LC_ADDRESS=C               LC_TELEPHONE=C             LC_MEASUREMENT=en_GB.UTF-8 LC_IDENTIFICATION=C       `

`attached base packages:`
`[1] stats     graphics  grDevices utils     datasets  methods   base     `

`other attached packages:`
`[1] h2owave_0.0.14`

`loaded via a namespace (and not attached):`
` [1] httr_1.4.2     compiler_4.0.4 magrittr_2.0.1 R6_2.5.0       tools_4.0.4    curl_4.3       tinytex_0.30   stringi_1.5.3 `
` [9] jsonlite_1.7.2 stringr_1.4.0  xfun_0.22  `

**Reply by ashrith** [reply]
@woobe thank you for the feedback. We will add more examples and you will see that soon.  The first set of examples were kept simple to assess the readability, brevity, and usability of the package and the examples.

**Reply by woobe** [reply]
I just installed the latest `h2owave_0.0.14.tar.gz` and reran all examples using wave server v0.13. All good!

**Reply by woobe** [reply]
Was the installation of the library easy? - **YES**

Did you understand each of the examples clearly? - **YES**

Was the code readable? - **YES**

Did code in the examples make sense to you as a R user? - **YES**

Was the code idiomatic? - **YES**

Were you able to execute the examples easily? **YES** (I would uncomment `library(h2owave)` in the first example `hello_world.R`

Was the code readable? - **YES**

Was the code brief? - **YES**

Was the code ergonomic? - **I would split the comments into a few lines if they exceed the 80 char count (standard R margin)**

Any other comments you would like to provide about the package? examples? experience? - **Let's create more examples together!**

**Reply by lo5** [reply]
Thanks for the review @woobe!

**Comment by brandenkmurray** [comment]
I'm having issues getting this to work on Mac. When I run one of the apps the wave server appears to be working as I can see stuff happening in terminal. For example, while `stat_small.R` runs I can see the following in terminal:

```
2021/03/26 09:53:38 * /demo {"d":[{"k":"crypto title","v":"LTC"},{"k":"crypto value","v":"131.654218069278"}]}
2021/03/26 09:53:42 * /demo {"d":[{"k":"crypto title","v":"LTC"},{"k":"crypto value","v":"375.661471097497"}]}
2021/03/26 09:53:45 * /demo {"d":[{"k":"crypto title","v":"BTC"},{"k":"crypto value","v":"95.5014126531314"}]}
2021/03/26 09:53:48 * /demo {"d":[{"k":"crypto title","v":"BTC"},{"k":"crypto value","v":"329.656990627758"}]}
2021/03/26 09:53:51 * /demo {"d":[{"k":"crypto title","v":"ETH"},{"k":"crypto value","v":"421.359045060351"}]}
2021/03/26 09:53:54 * /demo {"d":[{"k":"crypto title","v":"BTC"},{"k":"crypto value","v":"234.171027677134"}]}
2021/03/26 09:53:57 * /demo {"d":[{"k":"crypto title","v":"LTC"},{"k":"crypto value","v":"101.646449375199"}]}
2021/03/26 09:54:00 * /demo {"d":[{"k":"crypto title","v":"ZCH"},{"k":"crypto value","v":"303.540865081595"}]}
2021/03/26 09:54:03 * /demo {"d":[{"k":"crypto title","v":"LTC"},{"k":"crypto value","v":"432.879340720363"}]}
2021/03/26 09:54:06 * /demo {"d":[{"k":"crypto title","v":"ZCH"},{"k":"crypto value","v":"338.60062143812"}]}
2021/03/26 09:54:09 * /demo {"d":[{"k":"crypto title","v":"ZCH"},{"k":"crypto value","v":"4.90247914032079"}]}
2021/03/26 09:54:12 * /demo {"d":[{"k":"crypto title","v":"ETH"},{"k":"crypto value","v":"489.844166039256"}]}
```

However, when I navigate to [localhost:10101/demo](url) I just get a blank page. I've tried reloading, switching from Chrome to Safari, and changing `demo` to `demo2` in the code, but nothing has worked so far. I also originally had R 3.6.0 and then upgraded to 4.0.4 -- didn't work with either. Any guidance for where to go from here?



```
R version 4.0.4 (2021-02-15)
Platform: x86_64-apple-darwin17.0 (64-bit)
Running under: macOS Catalina 10.15.7

Matrix products: default
BLAS:   /System/Library/Frameworks/Accelerate.framework/Versions/A/Frameworks/vecLib.framework/Versions/A/libBLAS.dylib
LAPACK: /Library/Frameworks/R.framework/Versions/4.0/Resources/lib/libRlapack.dylib

locale:
[1] en_GB.UTF-8/en_GB.UTF-8/en_GB.UTF-8/C/en_GB.UTF-8/en_GB.UTF-8

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] h2owave_0.0.14

loaded via a namespace (and not attached):
[1] httr_1.4.2     compiler_4.0.4 magrittr_2.0.1 R6_2.5.0       tools_4.0.4    curl_4.3       stringi_1.5.3  jsonlite_1.7.2 stringr_1.4.0 
```

**Reply by ashrith** [reply]
@brandenkmurray Did you pick up the updated package in this discussion thread and install it? If not could you do that? The package sent in the email got a recent fix. Could you also share what version were you getting this problem? Was it 0.13? 
Pasting the link again here https://github.com/h2oai/wave/files/6201635/h2owave_0.0.14.tar.gz

**Reply by brandenkmurray** [reply]
Was the installation of the library easy?


Yes. My only suggestion is to make the instructions clearer. Instruction 3 is to download wave server, then Instruction 4  says to run `./waved -debug`. It might not be obvious to some that they need to extract the tar file and then navigate to the resulting directory before running the `./waved -debug` command. 

Did you understand each of the examples clearly?

		Yes.

Was the code readable?

	            Yes.


Did code in the examples make sense to you as a R user?

	      Yes.

Was the code idiomatic?

			Yes



Were you able to execute the examples easily?

	     Yes.

Was the code readable?

	                In `layout.R` line 24, I would add another indent so it’s clear that line is run in the inner for-loop. Everything else looks good to me.

Was the code brief?

	                Yes

Was the code ergonomic?

	               Yes

Any other comments you would like to provide about the package? examples? experience?


            I think an example that’s a little more complex would be nice. One that combines together everything in the current examples. Like the one we have in the image at https://h2oai.github.io/wave/

**Comment by grigory93** [comment]
1. Was the installation of the library easy? 
Yes

2. Did you understand each of the examples clearly?
Understood them clearly, but for people not familiar with R R6 OOP system this could be not straight forward to start. This is not to criticize R6 approach but rather to expect that not everyone is immediately familiar with it.

2.1. Was the code readable?
Yes, but  suggest more spacing between lines of code and more logical grouping of code segments based on structure, such as initialization/setup, layout, interaction, updating.

3. Did code in the examples make sense to you as a R user?
Yes, but, again, for those not familiar with R6 could be a learning curve.

Was the code idiomatic?
at the detail level Yes, at the higher app level No.

Were you able to execute the examples easily?
Yes

Was the code readable?
Yes

Was the code brief?
Yes, but examples were rather simple, so far.

Was the code ergonomic?
Yes, but again need to see more functional examples.

Any other comments you would like to provide about the package? examples? experience?
Having explicitly to place the code inside `while(TRUE)` loop (event loop I suppose) appears to be immature for a framework. There is hardly a case when this doesn't happen, unless it's a static image. So the questions are does the loop have to be explicit and why?

**Reply by lo5** [reply]
> Having explicitly to place the code inside while(TRUE) loop (event loop I suppose) appears to be immature for a framework.

@grigory93 They are not event loops. Those examples periodically update values on a page. In the real world you'd do this in response to some external event like a queue or cron.

**Reply by ashrith** [reply]
@grigory93 thanks for the response! 

>Did you understand each of the examples clearly?
Understood them clearly, but for people not familiar with R R6 OOP system this could be not straight forward to start. This is not to criticize R6 approach but rather to expect that not everyone is immediately familiar with it.

Could you please highlight which part of the code was not straight forward?  Any other information that you might think would help a user? 

>2.1. Was the code readable?
Yes, but suggest more spacing between lines of code and more logical grouping of code segments based on structure, such as initialization/setup, layout, interaction, updating.

The idea with the line-by-line code description is to tell the user in detail the goal of each line - which would tend to be a logical grouping.

For example:
```
# Combining 5 different crypto currency symbols into a vector.
crypto_name <- c("ETH","BTC","ZCH","LTC")
# Sample a single crypto currency symbol from the above created vector.
sample_crypto_name <- crypto_name[sample(1:length(crypto_name),1)]
# Generate a random value from an uniform distribution between 1 - 500.
sample_crypto_price <- runif(1,1,500)
# Generate a random decimal value from an uniform distribution between 0 - 1.
sample_crypto_price_percent_change <- runif(1,0,1)
# Check if a variable with the name "page_name" exists in an object page. This is a check to remove any previous page objects.
```
Here while reading thru the description the user is able to gather that this is __variable initialization phase__. We could make this explicit if it helps. 

>Was the code idiomatic?
at the detail level Yes, at the higher app level No.

> Was the code brief?
Yes, but examples were rather simple, so far.

The idea here is to keep the examples simple and show-by-code functional similarity to Python API without compromising on idiomatic ways of R.  Which is why the examples are not complex and do not lead to a complete application. 

Thank you.

**Reply by grigory93** [reply]
Hopefully, having a basic/minimal/trivial example that involves consuming and reacting to user input will add some clarity.

**Reply by grigory93** [reply]
I've never used R6 system before so I just expressed my reaction. For example, this code appears as a weird mix of legacy and new R6:
```
if("page_name" %in% names(page)) page$drop()
```

i'd say something like this would be more clear:
```
if (page$name() == "page_name") page$drop()

# or

if (page$get_name() == "page_name") page$drop()
```

**Reply by ashrith** [reply]
@grigory93, it is totally fine if R6 library is not familiar. What we are looking to find out is if this is completely alien or if it is in familiar territory.   The way to look at a R6 object is like accessing the elements in a list. 
For example, here we have a simple list (no R6 used here)
```
> x <- list()
> y <- function(...){print(...)}
> x$a <- 10
> x$b <- y
> x
$a
[1] 10

$b
function(val){print(val+1)}

> 
```
An implementation in R6 would be 
```
> library(R6)
> xx <- R6::R6Class("xx"
,public=
list(a=0
,printnum = function(){print(self$a)}))
```
Here notice that a similar list is created with the variable and function attached. 
Now, having said that, in this code

`if("page_name" %in% names(page)) page$drop()` 
is actually idiomatic R and not specific to R6. [Here we are using an `if()` condition in one line.](https://stackoverflow.com/questions/15586566/if-statement-in-r-can-only-have-one-line) 
As `names(page)` can give us a vector, using `%in%` we conduct a [(vector) value match.](https://stackoverflow.com/questions/42637099/difference-between-the-and-in-operators-in-r)
Hope this helps, please let me know if you have any questions.

**Reply by grigory93** [reply]
@ashrith what's next for R framework? would it be possible to test/ build whole apps with it soon/yet?

**Reply by ashrith** [reply]
Hi @grigory93 yes! Definitely. The first version that we are releasing is going to be dashboards. We are following it up with whole-build apps, that have similar interactivity as the python framework. Here is the link  for whats planned https://github.com/h2oai/wave/issues/15
